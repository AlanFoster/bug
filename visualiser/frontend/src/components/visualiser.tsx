import React, { useRef, useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Box, Button, Paper } from "@material-ui/core";
import { CodeEditor } from "./code-editor";
import { useHorizontalDrag } from "./use-horizontal-drag";
import { Memory } from "./memory";
import { Console } from "./console";
import { Resizer } from "./resizer";
import { defaultCode } from './default-code';
import clsx from "clsx";
import * as webassembly from "./webassembly";


const headingSizes = {
    height: 50,
};

const useStyles = makeStyles(theme => ({
    root: {

    },
    heading: {
        ...headingSizes,
        borderBottom: '1px solid rgba(0,0,0,0.2)',
        boxShadow: theme.shadows[1],
        padding: theme.spacing(1, 2),
    },
    content: {
        padding: theme.spacing(2),
        display: 'flex',
        height: `calc(100vh - ${headingSizes.height}px)`
    },
    panel: {
        display: 'flex',
        flexDirection: 'column'
    },
    secondaryPanel: {
        flex: 1
    },
    primaryChild: {
        flex: 1,
        overflow: 'scroll'
    },
    paper: {
        height: '100%',
        padding: theme.spacing(2)
    }
}));

const clamp = function (x: number, lowest: number, highest: number) {
    return Math.min(Math.max(lowest, x), highest);
};

const executeWebAssemblyText = async function (webAssemblyText: string) {
    const wasmBytes = await webassembly.compileText(webAssemblyText);

    let logs = [];
    var importObject = {
        "System::Output": {
            println: function (value) {
                logs = logs.concat(value);
            }
        }
    };

    const {instance} = await WebAssembly.instantiate(wasmBytes, importObject);
    instance.exports.Main();
    return {
        logs,
        memory: (
            instance.exports.memory
                ? instance.exports.memory.buffer
                : new Int8Array()
        )
    };
};

export const Visualiser = function () {
    const classes = useStyles({});
    const ref = useRef<Element>(null);
    const [code, setCode] = useState(defaultCode);
    const [executionResult, setExecutionResult] = useState({logs: [], memory: null});
    const {width: codePanelWidth, setDragging} = useHorizontalDrag(
        ref
    );

    const executeCode = async () => {
        try {
            const executionResult = await executeWebAssemblyText(code);
            setExecutionResult(executionResult);
        } catch (e) {
            setExecutionResult({ logs: [`Failure: ${e.stack}`], memory: null })
            console.error(e);
        }
    };

    useEffect(() => {
        executeCode();
    }, [])

    return (
        <Box className={classes.root} ref={ref}>
            <Box className={classes.heading}>
                <Button
                    color="primary"
                    variant="outlined"
                    onClick={executeCode}
                >
                    Run File
                </Button>
            </Box>

            <Box className={classes.content}>
                <Box
                    className={classes.panel}
                    width={
                        typeof codePanelWidth !== 'undefined'
                            ? `${clamp(codePanelWidth, 300, 1000)}px`
                            : '50%'
                    }
                >
                    <Paper className={clsx([classes.primaryChild, classes.paper])}>
                        <CodeEditor code={code} onValueChange={setCode}/>
                    </Paper>
                </Box>

                <Resizer
                    onMouseDown={() => setDragging(true)}
                />

                <Box
                    className={clsx([classes.panel, classes.secondaryPanel])}
                >
                    <Paper className={clsx([classes.primaryChild, classes.paper])}>
                        <Memory memory={executionResult.memory}/>
                    </Paper>
                    <Box>
                        <Paper className={classes.paper}>
                            <Console logs={executionResult.logs}/>
                        </Paper>
                    </Box>
                </Box>
            </Box>
        </Box>
    );
}
