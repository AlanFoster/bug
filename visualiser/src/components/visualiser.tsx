import React, { useRef } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Box } from "@material-ui/core";
import { CodeEditor } from "./code-editor";
import { useHorizontalDrag } from "./use-horizontal-drag";
import { Memory } from "./memory";
import { Console } from "./console";
import { Resizer } from "./resizer";
import { Panel } from "./panel";

const useStyles = makeStyles(theme => ({
    root: {
        display: 'flex',
        flexGrow: 1,
        height: '100vh'
    },
    runtimePanelWrapper: {
        flex: 1
    },
    primaryPanel: {
        flex: 1
    },
}));

const CodeActions = function () {
    return (
        <React.Fragment>
            Actions - Run / Upload
        </React.Fragment>
    )
};

export const Visualiser = function () {
    const classes = useStyles({});
    const ref = useRef<Element>(null);
    const {width: codePanelWidth, setDragging} = useHorizontalDrag(
        ref
    );

    return (
        <Box className={classes.root} ref={ref}>
            <Box
                width={
                    typeof codePanelWidth !== 'undefined'
                        ? `${Math.max(codePanelWidth, 300)}px`
                        : '50%'
                }
            >
                <Panel container>
                    <Panel item className={classes.primaryPanel}>
                        <CodeEditor/>
                    </Panel>
                    <Panel item>
                        <CodeActions/>
                    </Panel>
                </Panel>
            </Box>

            <Resizer
                onMouseDown={() => setDragging(true)}
            />

            <Box className={classes.runtimePanelWrapper}>
                <Panel container>
                    <Panel item className={classes.primaryPanel}>
                        <Memory/>
                    </Panel>
                    <Panel item>
                        <Console/>
                    </Panel>
                </Panel>
            </Box>
        </Box>
    );
}
