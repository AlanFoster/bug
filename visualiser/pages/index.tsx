import { Visualiser } from "../src/components/visualiser";

export default Visualiser;
