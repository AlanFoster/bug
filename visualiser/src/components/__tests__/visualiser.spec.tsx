import * as React from "react";
import renderer from "react-test-renderer"
import { Visualiser } from "../visualiser";

describe("Index", function () {
    it("renders successfully", function () {
        const tree = renderer.create(<Visualiser/>);

        expect(tree).toMatchSnapshot();
    });
});
