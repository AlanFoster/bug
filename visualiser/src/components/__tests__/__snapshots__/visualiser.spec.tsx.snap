// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index renders successfully 1`] = `
<div
  className="MuiBox-root MuiBox-root-5 makeStyles-root-1"
>
  <div
    className="MuiBox-root MuiBox-root-6"
  >
    <div
      className="MuiGrid-root makeStyles-root-7 MuiGrid-container MuiGrid-direction-xs-column"
    >
      <div
        className="MuiGrid-root makeStyles-primaryPanel-3 MuiGrid-item"
      >
        Code
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        Actions - Run / Upload
      </div>
    </div>
  </div>
  <div
    className="MuiBox-root MuiBox-root-113 makeStyles-root-111"
    onMouseDown={[Function]}
  >
    <div
      className="MuiBox-root MuiBox-root-114 makeStyles-handle-112"
    />
  </div>
  <div
    className="MuiBox-root MuiBox-root-115 makeStyles-runtimePanelWrapper-2"
  >
    <div
      className="MuiGrid-root makeStyles-root-7 MuiGrid-container MuiGrid-direction-xs-column"
    >
      <div
        className="MuiGrid-root makeStyles-primaryPanel-3 MuiGrid-item"
      >
        Memory
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        Console
      </div>
    </div>
  </div>
</div>
`;
